include/ginkgo/core/base/mpi.hpp

```diff
@@ -0,0 +1,654 @@
+
+    ~communicator();
+
+private:
+    MPI_Comm comm_;
+    int size_{};
+    int rank_{};
+    int local_rank_{};
+};
+
+
+/**
+ * A type helper which can be used to create MPI_Datatype from other types.
+ */
+class mpi_type {
+public:
+    mpi_type(const int count, MPI_Datatype& old);
+    ~mpi_type();
+    const MPI_Datatype& get() const { return this->type_; }
+
+private:
+    MPI_Datatype type_{};
+};
+
+
+/**
+ * This class wraps the MPI_Window class with RAII functionality. Different
+ * create and lock type methods are setup with enums.
+ *
+ * MPI_Window is primarily used for one sided communication and this class
+ * provides functionalities to fence, lock, unlock and flush the communication
+ * buffers.
+ */
+template <typename ValueType>
+class window {
+public:
+    enum class win_type { allocate = 1, create = 2, dynamic_create = 3 };
+    enum class lock_type { shared = 1, exclusive = 2 };
+
+    window() : window_(MPI_WIN_NULL) {}
+    window(window& other) = default;
+    window& operator=(const window& other) = default;
+    window(window&& other) = default;
+    window& operator=(window&& other) = default;
+
+    window(ValueType* base, unsigned int size,
+           std::shared_ptr<const communicator> comm,
+           const int disp_unit = sizeof(ValueType),
+           info input_info = info(MPI_INFO_NULL),
+           win_type create_type = win_type::create);
+
+    MPI_Win get() { return this->window_; }
+
+    void fence(int assert = 0);
+
+    void lock(int rank, int assert = 0, lock_type lock_t = lock_type::shared);
+
+    void unlock(int rank);
+
+    void lock_all(int assert = 0);
+
+    void unlock_all();
+
+    void flush(int rank);
+
+    void flush_local(int rank);
+
+    void flush_all();
+
+    void flush_all_local();
+
+    ~window();
+
+private:
+    MPI_Win window_;
+};
+
+
+/**
+ * This function is used to synchronize between the ranks of a given
+ * communicator.
+ *
+ * @param comm  the communicator
+ */
+void synchronize(const communicator& comm = communicator::get_comm_world());
+
+
+/**
+ * Allows a rank to wait on a particular request handle.
+ *
+ * @param req  The request to wait on.
+ * @param status  The status variable that can be queried.
+ */
+void wait(std::shared_ptr<request> req, std::shared_ptr<status> status = {});
+
+
+double get_walltime();
+
+
+/**
+ * Get the rank in the communicator of the calling process.
+ *
+ * @param comm  the communicator
+ */
+int get_my_rank(const communicator& comm = communicator::get_comm_world());
+
+
+/**
+ * Get the node local rank in the communicator of the calling process.
+ *
+ * @param comm  the communicator
+ */
+int get_local_rank(const communicator& comm = communicator::get_comm_world());
+
+
+/**
+ * Get the number of ranks in the communicator of the calling process.
+ *
+ * @param comm  the communicator
+ */
+int get_num_ranks(const communicator& comm = communicator::get_comm_world());
```

Comment:
Do we actually need these functions? The information is already cached in `communicator` and I don't think you can modify a comm object easily to make them invalid.